java/org/apache/catalina/webresources/AbstractSingleArchiveResourceSet.java

```diff
@@ -51,6 +57,10 @@ public AbstractSingleArchiveResourceSet(WebResourceRoot root, String webAppMount
         setBase(base);
         setInternalPath(internalPath);
 
+        if (isCacheArchiveEntries()) {
+            archiveEntries = loadArchiveEntries();
```

Comment:
do we know how the no-arg constructor is being used? if something was manually creating a `AbstractSingleArchiveResourceSet` by calling setters (some sort of bean operation?) we wouldn't populate it, and then it would always have an empty map.

Reply:
They indicate that the no-arg constructor is used by the "digester"...no idea what that is, but there's no setter method for `archiveEntries`

Reply:
right, im just thinking about if someone did
```
resource = new AbstractSingleArchiveResourceSet(); // some real impl extending this 
resource.setRoot(root);
resource.setWebAppMount(webAppMount);
resource.setBase(base);
resource.setInternalPath(internalPath);
resource.getResource("myresource")
```
it would always return null

Reply:
The digester seems to be some weird thingy for parsing in XML based rules.  I'm pretty confident we're not exposed, but I think it would be fine to simply add a null check an `archiveEntries` to `getArchiveEntries` and let it fall back to the original synchronized loading logic that follows in that method.

The good news is that the penalty would not be steep...we'd still only have to load the entries once and then when we have the map we'd never need to reload it due to our `gc()` modification in this commit.